### PR TITLE
Upgrade to Grain 0.4

### PR DESCRIPTION
**Pull request by phated**
Grain 0.4 converted all of the WASI functions from throwing exceptions to returning a `Result` type. This is a preliminary conversion from Grain 0.3.2 to 0.4.2.

There are a bunch of `TODO` comments asking what should happen if a Result is an `Err` type. I can make the changes if I know what should happen.

This version of Grain also added automatic code formatting, so a lot of the diff is that being applied. I also changed the graindoc style and added a make task to generate markdown documentation for everything in `lib`. (I'm bad with make, so I'm not sure this is the written the best way)

### CODE REVIEW

**Comment by phated** [comment]
Ref #1

File: Makefile
Comment (technosophos) [inline Makefile:32]: This is SO COOL.

File: fileserver.gr
Comment (technosophos) [inline fileserver.gr]: Here's what I am thinking -- and I am totally open to alternatives.

- Maybe if the `Content-type` header is concatenated to a string, and then written in one `File.fdWrite` call, we could trap any error and do a `fail`. That way, Wagi will notice this and send the client an HTTP 500.
- If the `pipe` call returns an Err variant but we have already sent the Content-Type header, then I think the best we can do is return. That will result in (at best) partial results or (at worst) an empty body.

As an alternative, if the `pipe` fails, we _can_ still `fail` at this point, because Wagi will not flush the headers until it gets the body (because currently it tries to send Content-Length, and in the future it will try to send the headers and the first chunk, I think). So in that case, even though we have sent the headers to Wagi, if we `fail`, Wagi could still send a 500 to the client.

File: fileserver.gr
Comment (technosophos) [inline fileserver.gr]: I think probably failing is best.

File: fileserver.gr
Comment (technosophos) [inline fileserver.gr]: Just thinking out loud here... and I may be totally misguided....

Maybe we should return an `Err` option as soon as an error is encountered, and otherwise keep recursing until the input has been exhausted, then return `Ok(len)` or something.

That way, we would be able to handle (somewhat gracefully) the situation where a `pipe` got interrupted. I'm not sure that in practice this helps us much at this point... but it feels sensible.

File: lib/stringutil.gr
Comment (technosophos) [inline lib/stringutil.gr:12]: This makes me happy... and also makes me feel slightly guilty because I had intended to contribute a `String.reverse`. 😁

**Review by technosophos — CHANGES_REQUESTED** [review]
Wow! Thank you so much! You totally went above and beyond!

I tried to answer your inline comments and at least give a few options. I'm still unsure about whether it's better to send back a partial response when a data read fails midway, or whether to fail and cause the runtime to generate a 500. I'd be happy to go whichever way you think is best.

Also, the formatter and doc engine are so great.

File: fileserver.gr
Comment (phated) [reply fileserver.gr]: I like it!

**Comment by phated** [comment]
I've not abandoned this, but I'm waiting for some bug fixes to land as Grain v0.4.3 before this should actually be merged. I'll do the updates once that is ready 🎉

**Comment by phated** [comment]
@technosophos I believe this has been updated successfully based on your suggestions! Let me know if you'd want anything to happen differently.

I also bumped Grain to 0.4.3, which we released today, that contained fixes for some outstanding bugs in our 0.4.x release cycle.

File: lib/stringutil.gr
Comment (radu-matei) [reply lib/stringutil.gr:12]: Not to worry, you still have `Array.sort` and `List.sort`.
https://github.com/grain-lang/grain/issues/125

**Review by radu-matei — APPROVED** [review]
Looks good to me, thanks!

(I really like the improvements in the standard library, they are awesome!)

**Comment by phated** [comment]
> (I really like the improvements in the standard library, they are awesome!)

Thanks! Updating this project has also given us some ideas for other stdlib functions (maybe `Result.testOk` and `Result.testErr` to do away with the `validateResult` function I wrote). And I think we want a nicer Env type to work with, like y'all implemented 🎉

**Pull request merged**
